In [39]:
import pandas as pd
from sklearn.model_selection import KFold, GridSearchCV
from sklearn.preprocessing import StandardScaler
from sklearn.metrics import classification_report, confusion_matrix
from sklearn.neighbors import KNeighborsClassifier

# Load the dataset
data_path = '/workspaces/Final-Year-Project/Cleaned Data/TrainTestData.csv'
df = pd.read_csv(data_path)

# Drop rows with NaN in any feature column
df = df.dropna()

# Define features and target
X = df.drop('Diabetes Status', axis=1)
y = df['Diabetes Status']

# Standardize the features (important for KNN)
scaler = StandardScaler()
X = scaler.fit_transform(X)

# Initialize the KNN model
knn = KNeighborsClassifier()

# Set up the hyperparameter grid for tuning
param_grid = {
    'n_neighbors': [3],  # Number of neighbors to consider
}

# Use GridSearchCV with KFold cross-validation
kf = KFold(n_splits=5, shuffle=True, random_state=42)
grid_search = GridSearchCV(knn, param_grid, scoring='accuracy', cv=kf, n_jobs=-1, verbose=1)

# Fit GridSearchCV on the data
grid_search.fit(X, y)

# Get the best model
best_knn = grid_search.best_estimator_

# Evaluate the model with the best parameters
y_pred = best_knn.predict(X)

# Print the best hyperparameters and the mean accuracy from cross-validation
print("Best Parameters for KNN:", grid_search.best_params_)
print("Mean Accuracy with Best Parameters:", grid_search.best_score_)

# Classification report and confusion matrix
print("KNN Classification Report:")
print(classification_report(y, y_pred))
print("Confusion Matrix:")
print(confusion_matrix(y, y_pred))


Fitting 5 folds for each of 1 candidates, totalling 5 fits
Best Parameters for KNN: {'n_neighbors': 3}
Mean Accuracy with Best Parameters: 0.6426910704381587
KNN Classification Report:
              precision    recall  f1-score   support

           0       0.79      0.75      0.77      1339
           1       0.80      0.84      0.82      1664

    accuracy                           0.80      3003
   macro avg       0.80      0.79      0.80      3003
weighted avg       0.80      0.80      0.80      3003

Confusion Matrix:
[[ 999  340]
 [ 261 1403]]


In [40]:
from sklearn.metrics import classification_report, confusion_matrix

# File path for validation data
valid_path = "/workspaces/Final-Year-Project/Cleaned Data/ValidationData.csv"

### 🎯 VALIDATION PHASE ###
# Load validation data
df_valid = pd.read_csv(valid_path)
df_valid = df_valid.dropna()

# Define features and target
X_valid = df_valid.drop('Diabetes Status', axis=1)
y_valid = df_valid['Diabetes Status']

# Standardize validation data (using the same scaler from training)
X_valid_scaled = scaler.transform(X_valid)

print(f"🧪 KNN - Validation dataset shape: {X_valid.shape}")

# Evaluate on validation set
y_pred_valid = best_knn.predict(X_valid_scaled)

# Classification Report
print(f"📌 KNN - Validation Classification Report:")
print(classification_report(y_valid, y_pred_valid))

# Confusion Matrix
print("🟦 KNN - Validation Confusion Matrix:")
print(confusion_matrix(y_valid, y_pred_valid))


🧪 KNN - Validation dataset shape: (335, 9)
📌 KNN - Validation Classification Report:
              precision    recall  f1-score   support

           0       0.66      0.68      0.67       145
           1       0.75      0.73      0.74       190

    accuracy                           0.71       335
   macro avg       0.70      0.70      0.70       335
weighted avg       0.71      0.71      0.71       335

🟦 KNN - Validation Confusion Matrix:
[[ 98  47]
 [ 51 139]]


best: nn = 5

🧪 KNN - Validation dataset shape: (335, 9)
📌 KNN - Validation Classification Report:
              precision    recall  f1-score   support

           0       0.72      0.68      0.70       145
           1       0.76      0.79      0.78       190

    accuracy                           0.74       335
   macro avg       0.74      0.74      0.74       335
weighted avg       0.74      0.74      0.74       335

🟦 KNN - Validation Confusion Matrix:
[[ 98  47]
 [ 39 151]]

nn = 7

🧪 KNN - Validation dataset shape: (335, 9)
📌 KNN - Validation Classification Report:
              precision    recall  f1-score   support

           0       0.68      0.66      0.67       145
           1       0.75      0.76      0.75       190

    accuracy                           0.72       335
   macro avg       0.71      0.71      0.71       335
weighted avg       0.72      0.72      0.72       335

🟦 KNN - Validation Confusion Matrix:
[[ 96  49]
 [ 46 144]]

nn = 3

🧪 KNN - Validation dataset shape: (335, 9)
📌 KNN - Validation Classification Report:
              precision    recall  f1-score   support

           0       0.66      0.68      0.67       145
           1       0.75      0.73      0.74       190

    accuracy                           0.71       335
   macro avg       0.70      0.70      0.70       335
weighted avg       0.71      0.71      0.71       335

🟦 KNN - Validation Confusion Matrix:
[[ 98  47]
 [ 51 139]]

nn = 9

🧪 KNN - Validation dataset shape: (335, 9)
📌 KNN - Validation Classification Report:
              precision    recall  f1-score   support

           0       0.67      0.62      0.64       145
           1       0.72      0.76      0.74       190

    accuracy                           0.70       335
   macro avg       0.70      0.69      0.69       335
weighted avg       0.70      0.70      0.70       335

🟦 KNN - Validation Confusion Matrix:
[[ 90  55]
 [ 45 145]]
